<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>ChAMP QC Pipeline for beta matrix — ChAMP_QC_Pipeline_Frome_Beta_Value • loonR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="ChAMP QC Pipeline for beta matrix — ChAMP_QC_Pipeline_Frome_Beta_Value" />
<meta property="og:description" content="ChAMP QC Pipeline for beta matrix" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



<!-- dependencies from examples -->


  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">loonR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.695</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ChAMP QC Pipeline for beta matrix</h1>
    <small class="dont-index">Source: <a href='https://github.com/ProfessionalFarmer/loonR/blob/master/R/Methylation.R'><code>R/Methylation.R</code></a></small>
    <div class="hidden name"><code>ChAMP_QC_Pipeline_Frome_Beta_Value.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>ChAMP QC Pipeline for beta matrix</p>
    </div>

    <pre class="usage"><span class='fu'>ChAMP_QC_Pipeline_Frome_Beta_Value</span><span class='op'>(</span>
  Sample.beta.df <span class='op'>=</span> <span class='cn'>NULL</span>,
  Sample.Group <span class='op'>=</span> <span class='st'>""</span>,
  Slide <span class='op'>=</span> <span class='st'>""</span>,
  arraytype <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"450K"</span>, <span class='st'>"EPIC"</span><span class='op'>)</span>,
  CpG.GUI <span class='op'>=</span> <span class='cn'>FALSE</span>,
  QC.GUI <span class='op'>=</span> <span class='cn'>FALSE</span>,
  combat <span class='op'>=</span> <span class='cn'>FALSE</span>,
  batchname <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Slide"</span><span class='op'>)</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Sample.beta.df</th>
      <td><p>Column is sample</p></td>
    </tr>
    <tr>
      <th>arraytype</th>
      <td><p>450K or EPIC</p></td>
    </tr>
    <tr>
      <th>CpG.GUI</th>
      <td><p>If use CpG.GUI</p></td>
    </tr>
    <tr>
      <th>combat</th>
      <td><p>If perform combat to remove batch effect</p></td>
    </tr>
    <tr>
      <th>batchname</th>
      <td><p>Batch variable name. Default is c("Slide")</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'></span>
<span class='r-in'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>ChAMP</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: minfi</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: BiocGenerics</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: parallel</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:parallel’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:loonR’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     plotPCA</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:dplyr’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     combine, intersect, setdiff, union</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:pROC’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     var</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     union, unique, unsplit, which.max, which.min</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: GenomicRanges</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: stats4</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: S4Vectors</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:dplyr’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     first, rename</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     expand.grid</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: IRanges</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘IRanges’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:dplyr’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     collapse, desc, slice</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: MatrixGenerics</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: matrixStats</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:dplyr’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     count</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: Biobase</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Welcome to Bioconductor</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘Biobase’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:MatrixGenerics’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     rowMedians</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     anyMissing, rowMedians</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: Biostrings</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: XVector</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘Biostrings’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:base’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     strsplit</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: bumphunter</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: foreach</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: iterators</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: locfit</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> locfit 1.5-9.4 	 2020-03-24</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Setting options('download.file.method.GEOquery'='auto')</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Setting options('GEOquery.inmemory.gpl'=FALSE)</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: ChAMPdata</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: DMRcate</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'minfi::getMeth' by 'bsseq::getMeth' when loading 'DMRcate'</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: Illumina450ProbeVariants.db</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: IlluminaHumanMethylationEPICmanifest</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: DT</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: RPMM</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Loading required package: cluster</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> No methods found in package 'RSQLite' for request: 'dbListFields' when loading 'lumi'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::summarise' by 'plotly::summarise' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::rename' by 'plotly::rename' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::arrange' by 'plotly::arrange' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::mutate' by 'plotly::mutate' when loading 'ChAMP'</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::is.discrete' by 'Hmisc::is.discrete' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plyr::summarize' by 'Hmisc::summarize' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plotly::subplot' by 'Hmisc::subplot' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'GenomicRanges::sort' by 'globaltest::sort' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'plotly::last_plot' by 'ggplot2::last_plot' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'globaltest::model.matrix' by 'stats::model.matrix' when loading 'ChAMP'</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>replacing previous import 'globaltest::p.adjust' by 'stats::p.adjust' when loading 'ChAMP'</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> &gt;&gt; Package version 2.20.1 loaded &lt;&lt;</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>        ___ _      _   __  __ ___ </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       / __| |_   /_\ |  \/  | _ \</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>      | (__| ' \ / _ \| |\/| |  _/</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       \___|_||_/_/ \_\_|  |_|_|  </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       ------------------------------</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     If you have any question or suggestion about ChAMP, please email to champ450k@gmail.com.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Thank you for citating ChAMP:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Yuan Tian, Tiffany J Morris, Amy P Webster, Zhen Yang, Stephan Beck, Andrew Feber, Andrew E Teschendorff; ChAMP: updated methylation analysis pipeline for Illumina BeadChips, Bioinformatics, btx513, https://doi.org/10.1093/bioinformatics/btx513</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>      --------------------------</span>
<span class='r-in'><span class='va'>testDir</span><span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>,package<span class='op'>=</span><span class='st'>"ChAMPdata"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>myImport</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/ChAMP/man/champ.import.html'>champ.import</a></span><span class='op'>(</span><span class='va'>testDir</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [===========================]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [&lt;&lt;&lt;&lt; ChAMP.IMPORT START &gt;&gt;&gt;&gt;&gt;]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> -----------------------------</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 1: Read PD Files Start ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   CSV Directory: /data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/lung_test_set.csv</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Find CSV Success</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Reading CSV File</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>incomplete final line found on '/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/lung_test_set.csv'</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Replace Sentrix_Position into Array</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Replace Sentrix_ID into Slide</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 1: Read PD file Done ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 2: Read IDAT files Start ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R03C02_Grn.idat ---- (1/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R05C02_Grn.idat ---- (2/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/9247377086_R01C01_Grn.idat ---- (3/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/9247377086_R02C01_Grn.idat ---- (4/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7766130112_R06C01_Grn.idat ---- (5/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7766130112_R01C02_Grn.idat ---- (6/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R01C01_Grn.idat ---- (7/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R01C02_Grn.idat ---- (8/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R03C02_Red.idat ---- (1/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R05C02_Red.idat ---- (2/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/9247377086_R01C01_Red.idat ---- (3/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/9247377086_R02C01_Red.idat ---- (4/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7766130112_R06C01_Red.idat ---- (5/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7766130112_R01C02_Red.idat ---- (6/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R01C01_Red.idat ---- (7/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Loading:/data0/Zhongxu/R/x86_64-pc-linux-gnu-library/4.0/ChAMPdata/extdata/7990895118_R01C02_Red.idat ---- (8/8)</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>truncating string with embedded nuls</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Extract Mean value for Green and Red Channel Success</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Your Red Green Channel contains 622399 probes.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 2: Read IDAT Files Done ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 3: Use Annotation Start ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Reading 450K Annotation &gt;&gt;</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Fetching NEGATIVE ControlProbe.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Totally, there are 613 control probes in Annotation.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Your data set contains 613 control probes.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Generating Meth and UnMeth Matrix</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Extracting Meth Matrix...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       Totally there are 485512 Meth probes in 450K Annotation.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       Your data set contains 485512 Meth probes.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Extracting UnMeth Matrix...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       Totally there are 485512 UnMeth probes in 450K Annotation.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>       Your data set contains 485512 UnMeth probes.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Generating beta Matrix</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Generating M Matrix</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Generating intensity Matrix</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Calculating Detect P value</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Counting Beads</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 3: Use Annotation Done ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [&lt;&lt;&lt;&lt;&lt; ChAMP.IMPORT END &gt;&gt;&gt;&gt;&gt;&gt;]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [===========================]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [You may want to process champ.filter() next.]</span>
<span class='r-in'></span>
<span class='r-in'><span class='va'>beta.df</span> <span class='op'>=</span> <span class='va'>myImport</span><span class='op'>$</span><span class='va'>beta</span></span>
<span class='r-in'><span class='va'>group</span> <span class='op'>=</span> <span class='va'>myImport</span><span class='op'>$</span><span class='va'>pd</span><span class='op'>$</span><span class='va'>Sample_Group</span></span>
<span class='r-in'><span class='fu'>loonR</span><span class='fu'>::</span><span class='fu'>ChAMP_QC_Pipeline_Frome_Beta_Value</span><span class='op'>(</span>Sample.beta.df<span class='op'>=</span><span class='va'>beta.df</span>, Sample.Group<span class='op'>=</span><span class='va'>group</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [===========================]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [&lt;&lt;&lt;&lt; ChAMP.FILTER START &gt;&gt;&gt;&gt;&gt;]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> -----------------------------</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> In New version ChAMP, champ.filter() function has been set to do filtering on the result of champ.import(). You can use champ.import() + champ.filter() to do Data Loading, or set "method" parameter in champ.load() as "ChAMP" to get the same effect.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> This function is provided for user need to do filtering on some beta (or M) matrix, which contained most filtering system in champ.load except beadcount. User need to input beta matrix, pd file themselves. If you want to do filterintg on detP matrix and Bead Count, you also need to input a detected P matrix and Bead Count information.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Note that if you want to filter more data matrix, say beta, M, intensity... please make sure they have exactly the same rownames and colnames.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 1:  Check Input Start ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   You have inputed beta for Analysis.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   pd file provided, checking if it's in accord with Data Matrix...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     pd file check success.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Parameter filterDetP is TRUE, checking if detP in accord with Data Matrix...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     !!! Parameter detP is not found, filterDetP is reset FALSE now.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Parameter filterBeads is TRUE, checking if beadcount in accord with Data Matrix...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     !!! Parameter beadcount is not found, filterBeads is reset FALSE now.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   parameter autoimpute is TRUE. Checking if the conditions are fulfilled...</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     !!! ProbeCutoff is 0, which means you have no needs to do imputation. autoimpute has been reset FALSE.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Checking Finished :filterMultiHit,filterSNPs,filterNoCG,filterXY would be done on beta.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 1: Check Input Done ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 2: Filtering Start &gt;&gt;</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Filtering NoCG Start</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Only Keep CpGs, removing 3091 probes from the analysis.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Filtering SNPs Start</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Using general 450K SNP list for filtering.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Filtering probes with SNPs as identified in Zhou's Nucleic Acids Research Paper 2016.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Removing 59901 probes from the analysis.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Filtering MultiHit Start</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Filtering probes that align to multiple locations as identified in Nordlund et al</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Removing 11 probes from the analysis.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Filtering XY Start</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     Filtering probes located on X,Y chromosome, removing 10028 probes from the analysis.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>   Updating PD file</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     filterDetP parameter is FALSE, so no Sample Would be removed.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [ Section 2: Filtering Done ]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>  All filterings are Done, now you have 412481 probes and 8 samples.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [&lt;&lt;&lt;&lt;&lt; ChAMP.FILTER END &gt;&gt;&gt;&gt;&gt;&gt;]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [===========================]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [You may want to process champ.QC() next.]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [===========================]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> [&lt;&lt;&lt; ChAMP.IMPUTE START &gt;&gt;&gt;&gt;]</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> -----------------------------</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> &lt;method&gt;:Combine. (Suitable for Large Data Set)</span>
<span class='r-err co'><span class='r-pr'>#&gt;</span> <span class='error'>Error in apply(beta, 2, function(x) length(which(is.na(x)))/length(x))</span> object 'myLoad' not found</span>
<span class='r-in'></span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Zhongxu.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


