<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ loonR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">loonR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.515</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="AllEqual.html">AllEqual()</a></code> </p>
        </td>
        <td><p>Check if all values in a vector are the same</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.GSEA.Compare.html">ClusterProfiler.GSEA.Compare()</a></code> </p>
        </td>
        <td><p>Perform GSE analysis across multiple group by cluster profiler</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.GSEA.ORA.customGS.Compare.html">ClusterProfiler.GSEA.ORA.customGS.Compare()</a></code> </p>
        </td>
        <td><p>Perform GSE analysis across multiple group by cluster profiler and user customed gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.GSEA.ORA.customGS.html">ClusterProfiler.GSEA.ORA.customGS()</a></code> </p>
        </td>
        <td><p>Perform analysis using user custom gene sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.GSEA.html">ClusterProfiler.GSEA()</a></code> </p>
        </td>
        <td><p>Perform GSEA analysis by cluster profiler</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.OverRepresentationTest.Compare.html">ClusterProfiler.OverRepresentationTest.Compare()</a></code> </p>
        </td>
        <td><p>Perform over-representation test across multiple group by cluster profiler</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ClusterProfiler.OverRepresentationTest.html">ClusterProfiler.OverRepresentationTest()</a></code> </p>
        </td>
        <td><p>Perform over-representation test by cluster profiler</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="DESeq2_differential.html">DESeq2_differential()</a></code> </p>
        </td>
        <td><p>Differential analysis by DESEQ2</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MA_plot.html">MA_plot()</a></code> </p>
        </td>
        <td><p>M-A plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MannWhitneyU_WilcoxonRankSumTest_differential.html">MannWhitneyU_WilcoxonRankSumTest_differential()</a></code> </p>
        </td>
        <td><p>Calculate p value by MannWhitneyU (WilcoxonRankSum) Test manually</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SNF_Similairity_Hist.html">SNF_Similairity_Hist()</a></code> </p>
        </td>
        <td><p>Similairity count</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build.best.random.forest.html">build.best.random.forest()</a></code> </p>
        </td>
        <td><p>Try different mtry and select the best fitted model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build.coxregression.model.html">build.coxregression.model()</a></code> </p>
        </td>
        <td><p>Build cox regression model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build.logistic.model.html">build.logistic.model()</a></code> </p>
        </td>
        <td><p>Build a logistic model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build.psm.regression.model.html">build.psm.regression.model()</a></code> </p>
        </td>
        <td><p>Build parametric survival model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="build.randomforest.model.html">build.randomforest.model()</a></code> </p>
        </td>
        <td><p>Build a random forests model</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare.GSE.HTSAnalyzer.html">compare.GSE.HTSAnalyzer()</a></code> </p>
        </td>
        <td><p>Perform GSEA across multiple groups (1 vs other)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compare.differential.analysis.html">compare.differential.analysis()</a></code> </p>
        </td>
        <td><p>Perform differential analysis (1 vs others)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="confidence_interval.html">confidence_interval()</a></code> </p>
        </td>
        <td><p>confidence interval</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="confusion_matrix.html">confusion_matrix()</a></code> </p>
        </td>
        <td><p>Get confusion matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertDfToNumeric.html">convertDfToNumeric()</a></code> </p>
        </td>
        <td><p>Convert data frame to numeric</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cross.validation.html">cross.validation()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="decisionCurveAnalysis.html">decisionCurveAnalysis()</a></code> </p>
        </td>
        <td><p>Decision curve analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="determineClusterNumber.html">determineClusterNumber()</a></code> </p>
        </td>
        <td><p>Determine the best cluster numbers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="download.geo.dataset.html">download.geo.dataset()</a></code> </p>
        </td>
        <td><p>Download dataset from GEO by accession ID and platform</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="draw.expression.dotplot.html">draw.expression.dotplot()</a></code> </p>
        </td>
        <td><p>Draw expression by dotplot
Need to be update</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="draw.mutation.type.html">draw.mutation.type()</a></code> </p>
        </td>
        <td><p>Count each event type and draw pie chart</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="draw.mutation.type.barplot.html">draw.mutation.type.barplot()</a></code> </p>
        </td>
        <td><p>Draw bar plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="draw.target.coverage.plot.html">draw.target.coverage.plot()</a></code> </p>
        </td>
        <td><p>Draw target coverage plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="drawScatter.html">drawScatter()</a></code> </p>
        </td>
        <td><p>Draw scatter plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="draw_confusion_matrix.html">draw_confusion_matrix()</a></code> </p>
        </td>
        <td><p>Confusion matrix visulization</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl_EntrezID.html">ensembl_EntrezID()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="export2ppt.html">export2ppt()</a></code> </p>
        </td>
        <td><p>Export figure to PPT
https://github.com/tomwenseleers/export</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extractSignaturePercent.html">extractSignaturePercent()</a></code> </p>
        </td>
        <td><p>Estimate signature percents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature.selection.4givenMLAlgorithm.html">feature.selection.4givenMLAlgorithm()</a></code> </p>
        </td>
        <td><p>Variable Importance from a specific Machine Learning Algorithms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature.selection.RFE.html">feature.selection.RFE()</a></code> </p>
        </td>
        <td><p>Recursive feature elimnation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="feature.selection.boruta.html">feature.selection.boruta()</a></code> </p>
        </td>
        <td><p>Feature selection by Boruta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterGenes.html">filterGenes()</a></code> </p>
        </td>
        <td><p>Filter genes by kinks of criteria</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findSomaticSignature.html">findSomaticSignature()</a></code> </p>
        </td>
        <td><p>Identify signature from sample data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="forest_plot.html">forest_plot()</a></code> </p>
        </td>
        <td><p>Forest plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fpkm2tpm.html">fpkm2tpm()</a></code> </p>
        </td>
        <td><p>Convert FPKM data fram to TPM</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gapStat.html">gapStat()</a></code> </p>
        </td>
        <td><p>Gap statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generate.sequencial.color.html">generate.sequencial.color()</a></code> </p>
        </td>
        <td><p>Generate sequencial color</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="generateCombinations.html">generateCombinations()</a></code> </p>
        </td>
        <td><p>Gnenerate combinations with specified size</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.AUC.html">get.AUC()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.TCGA.RNA.FPKM.html">get.TCGA.RNA.FPKM()</a></code> </p>
        </td>
        <td><p>Get TCGA RNA expression (RPKM) data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.TCGA.miRNAExpression.html">get.TCGA.miRNAExpression()</a></code> </p>
        </td>
        <td><p>Get TCGA miRNA expression data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.YoudenIndex.html">get.YoudenIndex()</a></code> </p>
        </td>
        <td><p>Get Youden index value</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.ggsci.color.html">get.ggsci.color()</a></code> </p>
        </td>
        <td><p>Get colors by ggsci</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.mostDistint.color.palette.html">get.mostDistint.color.palette()</a></code> </p>
        </td>
        <td><p>Generate a most distinctive color palatte</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.palette.color.html">get.palette.color()</a></code> </p>
        </td>
        <td><p>Use ggpubr function to get all kinds of color palette</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get.sample.project.infomap.html">get.sample.project.infomap()</a></code> </p>
        </td>
        <td><p>Get sample barcode and TCGA project map information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFastaSummary.html">getFastaSummary()</a></code> </p>
        </td>
        <td><p>Obtain sequence length from fasta file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGeneRelatedTranscripts.Gff.html">getGeneRelatedTranscripts.Gff()</a></code> </p>
        </td>
        <td><p>Retrieve all transcripts related to target genes from gtf/gff file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getOneRoundCVRes.html">getOneRoundCVRes()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProgramName.html">getProgramName()</a></code> </p>
        </td>
        <td><p>Get current running program name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPromoterRegions.html">getPromoterRegions()</a></code> </p>
        </td>
        <td><p>Get promoters by GenomicFeatures package. Region needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTranscriptLength.html">getTranscriptLength()</a></code> </p>
        </td>
        <td><p>Obtain transcript length from gff file or GRanges object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTranscriptProteinSequence.html">getTranscriptProteinSequence()</a></code> </p>
        </td>
        <td><p>Retrieve protein sequence by biomaRt</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_full_mapping_table.html">get_full_mapping_table()</a></code> </p>
        </td>
        <td><p>Retrieve full ID mapping table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_performance.html">get_performance()</a></code> </p>
        </td>
        <td><p>Title Get model performace, sensitivity, sepcificity and others</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="heatmap.annotation.html">heatmap.annotation()</a></code> </p>
        </td>
        <td><p>Only draw annotation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="heatmap.with.lgfold.riskpro.html">heatmap.with.lgfold.riskpro()</a></code> </p>
        </td>
        <td><p>Plot heatmap with log 2 fold change and risk probability information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hello.html">hello()</a></code> </p>
        </td>
        <td><p>Hello, World!</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hyperGeoTest.html">hyperGeoTest()</a></code> </p>
        </td>
        <td><p>Perform hypergeometric test</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="id_mapping.html">id_mapping()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="iflog2.html">iflog2()</a></code> </p>
        </td>
        <td><p>If perform log2 transformation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.nan.data.frame.html">is.nan(<i>&lt;data.frame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Generic is.nan function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lasso.cv.select.feature.html">lasso.cv.select.feature()</a></code> </p>
        </td>
        <td><p>Perform multiple round lasso to select stable feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lasso.select.feature.html">lasso.select.feature()</a></code> </p>
        </td>
        <td><p>Feature selection by lasso</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lasso_best_lamda.html">lasso_best_lamda()</a></code> </p>
        </td>
        <td><p>Get best lamda by perform multiple round lasso-sv
https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html#log
https://stackoverflow.com/questions/62170965/looping-cv-glmnet-and-get-the-best-coefficients</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="limma.differential.cv.selection.html">limma.differential.cv.selection()</a></code> </p>
        </td>
        <td><p>Perform multiple rounds differential analysis to select stable feature</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="limma_differential.html">limma_differential()</a></code> </p>
        </td>
        <td><p>Differential analysis by LIMMA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load.rsem.matrix.html">load.rsem.matrix()</a></code> </p>
        </td>
        <td><p>Load quantification result from RSEM by tximport</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load.salmon.matrix.html">load.salmon.matrix()</a></code> </p>
        </td>
        <td><p>Read Salmon output</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadGff.html">loadGff()</a></code> </p>
        </td>
        <td><p>Convert gff file to data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadMethylationArraryData.html">loadMethylationArraryData()</a></code> </p>
        </td>
        <td><p>Load methylation arrary from directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="log2dfQuantileNormalization.html">log2dfQuantileNormalization()</a></code> </p>
        </td>
        <td><p>Quantile normalization for log2 data frame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logit2prob.html">logit2prob()</a></code> </p>
        </td>
        <td><p>Convert logti to probability</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loo.cv.html">loo.cv()</a></code> </p>
        </td>
        <td><p>Leave one out cross validation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loo.cv.cox.html">loo.cv.cox()</a></code> </p>
        </td>
        <td><p>Leave one out cross validation for cox regression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="microarray_limma_differential.html">microarray_limma_differential()</a></code> </p>
        </td>
        <td><p>GEO dataset analysis by limma.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multi_roc_with_ci.html">multi_roc_with_ci()</a></code> </p>
        </td>
        <td><p>Plot multiple ROCs in one figure</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiplePlotPredictedPro.html">multiplePlotPredictedPro()</a></code> </p>
        </td>
        <td><p>plot predictive probability distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multivariate_cox.html">multivariate_cox()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multivariate_or.html">multivariate_or()</a></code> </p>
        </td>
        <td><p>Variate logistic analysis
Row: sample, Column: gene expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nomogram.plot.html">nomogram.plot()</a></code> </p>
        </td>
        <td><p>Nomogram plot by rms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="normalize_miRNACount_toCPM.html">normalize_miRNACount_toCPM()</a></code> </p>
        </td>
        <td><p>Normalize miRNA counts to CPM</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="oneway.anova.html">oneway.anova()</a></code> </p>
        </td>
        <td><p>Kruskal Wallis Test One Way Anova by Ranks or F test</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBarWithErr.html">plotBarWithErr()</a></code> </p>
        </td>
        <td><p>Plot bar with mean and standard error</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotClevelandDot.html">plotClevelandDot()</a></code> </p>
        </td>
        <td><p>Draw Celventland dot plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotJitterBoxplot.html">plotJitterBoxplot()</a></code> </p>
        </td>
        <td><p>Boxplot with jitter</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPCA.html">plotPCA()</a></code> </p>
        </td>
        <td><p>PCA plot
Please note: scale will be performed autmaticly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPie.html">plotPie()</a></code> </p>
        </td>
        <td><p>Count each event type and draw pie chart</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotSilhouette.html">plotSilhouette()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUpset.html">plotUpset()</a></code> </p>
        </td>
        <td><p>Upset plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVenn.html">plotVenn()</a></code> </p>
        </td>
        <td><p>Plot venn diagram by ggvenn</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVennLegacy.html">plotVennLegacy()</a></code> </p>
        </td>
        <td><p>Plot venn diagram</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_miRCorrelation.html">plot_miRCorrelation()</a></code> </p>
        </td>
        <td><p>Plot miR's correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot_waterfall.html">plot_waterfall()</a></code> </p>
        </td>
        <td><p>Waterfall plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read_mirdeep2_result.html">read_mirdeep2_result()</a></code> </p>
        </td>
        <td><p>Load miRDeep2 expression matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reduce_join_list.html">reduce_join_list()</a></code> </p>
        </td>
        <td><p>Join a table list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="refSeq_Ensembl.html">refSeq_Ensembl()</a></code> </p>
        </td>
        <td><p>Convert RefSeq ID to Ensembl ID by biomaRt</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="remove.verylong.isoform.gff.html">remove.verylong.isoform.gff()</a></code> </p>
        </td>
        <td><p>Remove isforms with extremly long interval (isoform.End - isoform.Start)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="riskCalibrationPlot.default.html">riskCalibrationPlot.default()</a></code> </p>
        </td>
        <td><p>#' riskCalibrationPlot
#'
#' @param x
#' @param ...
#'
#' @return
#' @export
#'
#' @examples
riskCalibrationPlot&lt;-function(group, pred, rms.method = FALSE, title = "Calibration plot", show.oberved.ci = FALSE,  bins = 10, color="npg", show.group = FALSE, ticks.unit=0.25, full.range=TRUE, ...) Generics function

UseMethod('riskCalibrationPlot')

Calibration plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="riskCalibrationPlot.lrm.html">riskCalibrationPlot.lrm()</a></code> </p>
        </td>
        <td><p>Calibration plot by rms.calibrate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rms_validate.html">rms_validate()</a></code> </p>
        </td>
        <td><p>Resampling Validation of a Fitted Model's Indexes of Fit</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="roc_with_ci.html">roc_with_ci()</a></code> </p>
        </td>
        <td><p>plot ROC with CI. Stratified bootstrap 2000 times</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_SNF.html">run_SNF()</a></code> </p>
        </td>
        <td><p>Run similarity network fusion</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="scaleDF.html">scaleDF()</a></code> </p>
        </td>
        <td><p>Scale a data.frame by column or row</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="show_hcluster.html">show_hcluster()</a></code> </p>
        </td>
        <td><p>Perform hclustering analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splitDataByGroup.html">splitDataByGroup()</a></code> </p>
        </td>
        <td><p>split data frame by group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splitGroupByCutoff.html">splitGroupByCutoff()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="splitSampleByGroup.html">splitSampleByGroup()</a></code> </p>
        </td>
        <td><p>Split sample by group into traning and validation set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="suppa.event.pie.html">suppa.event.pie()</a></code> </p>
        </td>
        <td><p>Count each event type and draw pie chart</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="suppa.get.final.table.html">suppa.get.final.table()</a></code> </p>
        </td>
        <td><p>Combine dpis, psi and tpm file together
I modified the code from https://github.com/comprna/SUPPA/blob/master/scripts/Volcano_MA_plot.R</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="survivaly_analysis.html">survivaly_analysis()</a></code> </p>
        </td>
        <td><p>Survival analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tanimoto_distance.html">tanimoto_distance()</a></code> </p>
        </td>
        <td><p>Calculate tanimoto distance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcgabiolinks.get.RNA.expression.log2tpm.html">tcgabiolinks.get.RNA.expression.log2tpm()</a></code> </p>
        </td>
        <td><p>Download RAN expression by TCGAbiolinks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tcgabiolinks.get.junction.coverage.html">tcgabiolinks.get.junction.coverage()</a></code> </p>
        </td>
        <td><p>Download junction data by TCGAbiolinks. Pls note, V2 tag and Illumina HiSeq platform were selected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="thisPath.html">thisPath()</a></code> </p>
        </td>
        <td><p>Get current directory path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="time_serials_Cindex.html">time_serials_Cindex()</a></code> </p>
        </td>
        <td><p>Draw C index across times</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ttest_differential.html">ttest_differential()</a></code> </p>
        </td>
        <td><p>Calculate p value by t.test manually</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unique_gene_expression.html">unique_gene_expression()</a></code> </p>
        </td>
        <td><p>Title Remove redundant gene expression data, and select the maximum one</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="univariate_cox.html">univariate_cox()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="univariate_cox_sthda.html">univariate_cox_sthda()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="univariate_or.html">univariate_or()</a></code> </p>
        </td>
        <td><p>Variate logistic analysis
Row: sample, Column: gene expression
score E.g.: Gene or miRNA expression, or risk score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="volcano_plot.html">volcano_plot()</a></code> </p>
        </td>
        <td><p>Volcano Plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeGff.html">writeGff()</a></code> </p>
        </td>
        <td><p>Write to gff file</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Zhongxu.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


